packages/flutter/lib/src/foundation/tree_diagnostics_mixin.dart

```diff
@@ -40,8 +1469,47 @@ String describeEnum(Object enumEntry) {
   return description.substring(indexOfDot + 1);
 }
 
-/// A mixin that helps dump string representations of trees.
-abstract class TreeDiagnosticsMixin {
+/// Returns a lower case hyphen separated version of a camel case name.
+///
+/// For example:
+///
+/// ```dart
+///
+/// main() {
+///   assert(toHyphenedName('deferToChild') == 'defer-to-child');
+///   assert(toHyphenedName('Monday') == 'monday');
+///   assert(toHyphenedName('monday') == 'monday');
+/// }
+/// ```
+String toHyphenedName(String word) {
+  final String lowerWord = word.toLowerCase();
+  if (word == lowerWord)
+    return word;
+
+  final StringBuffer sb = new StringBuffer();
+  for (int i = 0; i < word.length; i++) {
+    final String lower = lowerWord[i];
+    if (word[i] != lower && i > 0)
+      sb.write('-');
+    sb.write(lower);
+  }
+  return sb.toString();
+}
+
+/// An interface for classes providing string and [DiagnosticNode] debug
+/// representations of objects.
+///
+/// See also:
+///
+///  * [TreeDiagnosticsMixin]
+abstract class TreeDiagnostics {
+  String toStringDeep([String prefixLineOne = '', String prefixOtherLines = '']);
+  DiagnosticsNode toDiagnosticsNode({ String name, DiagnosticsTreeStyle style });
+}
+
+/// A mixin that helps dump string and [DiagnosticsNode] representations of
+/// trees.
```

Comment:
This documentation (or the docs for TreeDiagnostics) should include several paragraphs explaining how the whole system works.